src/plugins/searchDecorations.js

```diff
@@ -0,0 +1,100 @@
+import { Plugin, PluginKey } from '@tiptap/pm/state'
+import { Decoration, DecorationSet } from '@tiptap/pm/view'
+import { searchQueryPluginKey } from './searchQuery.js'
+
+/**
+ * Search decorations ProseMirror plugin
+ * Handles highlighting search matches for the search TipTap extension
+ *
+ * @return {Plugin<DecorationSet>}
+ */
+export default function searchDecorations() {
+	return new Plugin({
+		key: new PluginKey('searchDecorations'),
+		state: {
+			init(_, { doc }) {
+				const searchResults = runSearch(doc, '')
+				return highlightResults(doc, searchResults)
+			},
+			apply(tr, _, oldState, newState) {
+				const { query: oldQuery } = searchQueryPluginKey.getState(oldState)
+				const { query: newQuery } = searchQueryPluginKey.getState(newState)
+
+				let query = ''
+
+				if (tr.docChanged || (newQuery !== oldQuery)) {
+					query = newQuery
+				} else {
+					query = oldQuery
```

Comment:
I think you can just return `oldState` here. The doc did not change and the query also stayed the same.

That way you also don't need the `query` variable and can just use `newQuery`.

Reply:
Oh... the way this is defined in prosemirror is slightly confusing. I think you will need to use all four args of the function:
```js
apply(tr, value, oldState, newState) {
	...
	if (tr.docChanged || (newQuery !== oldQuery)) {
		...
	} else {
		return value // same as before
	}
```

Reply:
So, I think you mean something like this?

```js
apply(tr, value, oldState, newState) {
    const { query: oldQuery } = searchQueryPluginKey.getState(oldState)
    const { query: newQuery } = searchQueryPluginKey.getState(newState)

    if (tr.docChanged || (newQuery !== oldQuery)) {
        const searchResults = runSearch(tr.doc, newQuery)
        return highlightResults(tr.doc, searchResults)
    } else {
        return value
    }
},
```